### PR TITLE
Dreambooth text to image finetuning notebook examples

### PR DESCRIPTION
**Pull request by gauravrajguru**
# Description

# Checklist

- [x] I have read the contribution guidelines.
  - [General](https://github.com/Azure/azureml-examples/blob/main/CONTRIBUTING.md)
  - [SDK](https://github.com/Azure/azureml-examples/blob/main/sdk/CONTRIBUTING.md)
  - [CLI](https://github.com/Azure/azureml-examples/blob/main/cli/CONTRIBUTING.md)
- [x] I have coordinated with the docs team (mldocs@microsoft.com) if this PR deletes files or changes any file names or file extensions.
- [x] Pull request includes test coverage for the included changes.
- [x] This notebook or file is added to the [CODEOWNERS](https://github.com/Azure/azureml-examples/blob/main/.github/CODEOWNERS) file, pointing to the author or the author's team.

### CODE REVIEW

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+    "\n",
+    "### Model\n",
+    "We will use the `runwayml-stable-diffusion-v1-5` model in this notebook. If you need to fine tune a model that is available on HuggingFace, but not available in `azureml` system registry, you can either register the model and use the registered model or use the `model_name` parameter to instruct the components to pull the model directly from HuggingFace.\n",
+    "\n",
+    "### Outline\n",
+    "1. Install dependencies\n",
+    "2. Setup pre-requisites such as compute\n",
+    "3. Pick a model to fine tune\n",
+    "4. Prepare dataset for finetuning the model\n",
+    "5. Submit the fine tuning job using transformers specific image-classification component\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:24]:
```suggestion
    "5. Submit the fine tuning job using transformers specific text-to-image-generation component\n",
```

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 2]
```diff
@@ -0,0 +1,704 @@
...
+    "6. Review training and evaluation metrics\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:25]: Do we have the metrics for this?

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 3]
```diff
@@ -0,0 +1,704 @@
...
+    "We will use the `runwayml-stable-diffusion-v1-5` model in this notebook. If you need to fine tune a model that is available on HuggingFace, but not available in `azureml` system registry, you can either register the model and use the registered model or use the `model_name` parameter to instruct the components to pull the model directly from HuggingFace.\n",
+    "\n",
+    "Currently following models are supported:\n",
+    "\n",
+    "| Model Name | Source |\n",
+    "| ------ | ---------- |\n",
+    "| [runwayml-stable-diffusion-v1-5](https://ml.azure.com/registries/azureml/models/runwayml-stable-diffusion-v1-5/version/8) | azureml registry |\n",
+    "| [stabilityai-stable-diffusion-2-1](https://ml.azure.com/registries/azureml/models/stabilityai-stable-diffusion-2-1/version/8) | azureml registry |\n",
+    "| [compvis-stable-diffusion-v1-4](https://ml.azure.com/registries/azureml/models/compvis-stable-diffusion-v1-4/version/8) | azureml registry |\n",
+    "| [Text to Image models from Huggingface's Transformer library](https://huggingface.co/models?pipeline_tag=text-to-image&library=transformers)| HuggingFace |"
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:158]: deci diffusion model works as well. Please add that to the list. 
Sample run: https://ml.azure.com/runs/8d654f9f-1ac1-4398-9ad1-ea9f4e788b26?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/grajguru-rg/providers/Microsoft.MachineLearningServices/workspaces/grajguru-ws&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+    "print(\"\")\n",
+    "print(\"Path to folder in Blob Storage:\")\n",
+    "print(instance_data_uri_folder.path)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 5. Submit the fine tuning job using `transformers_text_to_image_dreambooth_finetuning_pipeline` component\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:263]:
```suggestion
    "### 5. Submit the fine tuning job using `diffusers_text_to_image_dreambooth_finetuning_pipeline` component\n",
```

I think better to call it "diffusers" instead of transformers.

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 5]
```diff
@@ -0,0 +1,704 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "FINETUNE_PIPELINE_COMPONENT_NAME = \"transformers_text_to_image_dreambooth_finetuning_pipeline\"\n",
+    "# Gaurav\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:282]: You want to remove this commented part?

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 6]
```diff
@@ -0,0 +1,704 @@
...
+    "#### 5.1 Get pipeline component"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "FINETUNE_PIPELINE_COMPONENT_NAME = \"transformers_text_to_image_dreambooth_finetuning_pipeline\"\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:281]:
```suggestion
    "FINETUNE_PIPELINE_COMPONENT_NAME = \"diffusers_text_to_image_dreambooth_finetuning_pipeline\"\n",
```

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 7]
```diff
@@ -0,0 +1,704 @@
...
+    "pipeline_component_transformers_func = registry_ml_client.components.get(\n",
+    "    name=FINETUNE_PIPELINE_COMPONENT_NAME, label=\"latest\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 5.2 Create arguments to be passed to `transformers_text_to_image_dreambooth_finetuning_pipeline` component\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:296]:
```suggestion
    "#### 5.2 Create arguments to be passed to `diffusers_text_to_image_dreambooth_finetuning_pipeline` component\n",
```

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 8]
```diff
@@ -0,0 +1,704 @@
...
+    "\n",
+    "The `transformers_text_to_image_dreambooth_finetuning_pipeline` component consists of model selection and finetuning components.\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:298]:
```suggestion
    "The `difusers_text_to_image_dreambooth_finetuning_pipeline` component consists of model selection and finetuning components.\n",
```

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 9]
```diff
@@ -0,0 +1,704 @@
...
+    "    json.dump(request_json, request_file)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "online_endpoint_name = \"text-to-image-dog-04291630\"\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:635]: Do you want to clear it up? I believe this was for testing?

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 10]
```diff
@@ -0,0 +1,704 @@
...
+    "    \"prior_generation_precision\": \"fp32\",\n",
+    "    \"prior_loss_weight\": 1.0,\n",
+    "    \"sample_batch_size\": 2,  # Number of samples to generate class images in each batch.\n",
+    "    # # Lora parameters\n",
+    "    # # LoRA reduces the number of trainable parameters by learning pairs of rank-decompostion matrices while freezing the original weights. This vastly reduces the storage requirement for large models adapted to specific tasks and enables efficient task-switching during deployment all without introducing inference latency. LoRA also outperforms several other adaptation methods including adapter, prefix-tuning, and fine-tuning.\n",
+    "    \"apply_lora\": True,\n",
+    "    # \"lora_alpha\": 128,\n",
+    "    \"lora_r\": 8,\n",
+    "    # \"lora_dropout\": 0.0,\n",
+    "    # \"tokenizer_max_length\": 77,\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:334]: Let's add the meaning of this one?

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 11]
```diff
@@ -0,0 +1,704 @@
...
+    "    # \"extra_noise_scheduler_args\": \"clip_sample_range=1.0; clip_sample=True\" # Optional additional arguments that are supplied to noise scheduler. The arguments should be semi-colon separated key value pairs and should be enclosed in double quotes.\n",
+    "    # \"offset_noise\": False\n",
+    "    # # Training related\n",
+    "    \"num_validation_images\": 3,  # Number of images to generate using instance_prompt. Images are stored in the output/checkpoint-* directories. Please note that this will increase the training time.\n",
+    "    \"number_of_workers\": 3,\n",
+    "    \"number_of_epochs\": 15,\n",
+    "    \"max_steps\": -1,\n",
+    "    \"training_batch_size\": 3,\n",
+    "    \"auto_find_batch_size\": False,\n",
+    "    \"learning_rate\": 1e-4,\n",
```
Comment (rjaincc) [inline sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:358]: Let's add a comment on learning rate being lower is lora is disabled

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb (hunk 7 above)
Comment (rjaincc) [reply sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:291]: Changed

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb (hunk 5 above)
Comment (rjaincc) [reply sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb]: Removed

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb [hunk 12]
```diff
@@ -0,0 +1,704 @@
...
+    "print(instance_data_uri_folder.path)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 5. Submit the fine tuning job using `transformers_text_to_image_dreambooth_finetuning_pipeline` component\n",
+    " \n",
+    "Create the job that uses the `transformers_text_to_image_dreambooth_finetuning_pipeline` component for stable-diffusion-text-to-image task. Learn more in 5.2 about all the parameters supported for fine tuning."
```
[the remark this answers is not in the document]
Comment (rjaincc) [reply sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:264]: Changed

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb (hunk 4 above)
Comment (rjaincc) [reply sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:262]: Changed

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb (hunk 3 above)
Comment (rjaincc) [reply sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:158]: Added

File: sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb (hunk 1 above)
Comment (rjaincc) [reply sdk/python/foundation-models/system/finetune/text-to-image/hftransformers-dreambooth-dog-text-to-image.ipynb:24]: Changed

File: sdk/python/foundation-models/system/finetune/text-to-image/diffusers-dreambooth-dog-text-to-image.ipynb [hunk 13]
```diff
@@ -0,0 +1,815 @@
...
+    "pipeline_component_transformers_func = registry_ml_client.components.get(\n",
+    "    name=FINETUNE_PIPELINE_COMPONENT_NAME, label=\"latest\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### 5.2 Create arguments to be passed to `diffusers_text_to_image_dreambooth_finetuning_pipeline` component\n",
```
Comment (shubhamiit) [inline sdk/python/foundation-models/system/finetune/text-to-image/diffusers-dreambooth-dog-text-to-image.ipynb:292]: There is a clear gap in understanding the variables like instance_prompt, class_prompt. Either we should definitions for the same or provide a hugging face link.

**Review by shubhamiit — APPROVED** [review]

**Review by rjaincc — APPROVED** [review]
